File: Python/linkedlist/singlylinkedlist.py
class SinglyNode:
	def __init__(self, data):
		self.data = data
		self.next = None


class SinglyLinkedList:
	def __init__(self, elems = None):
		self._head = None
		self._tail = None
		self._length = 0
		if elems:
			for e in elems:
				self.add_last(e)
	
	def __len__(self):
		return self._length
	
	def __getitem__(self, index):
		return self.get(index)
	
	def __setitem__(self, index, e):
		self.set(index, e)
	
	def __contains__(self, e):
		return self.contains(e)
	
	def empty(self):
		return self._length == 0
	
	@property
	def first(self):
		return getattr(self._head, 'data', None)
	
	@property 
	def last(self):
		return getattr(self._tail, 'data', None)
	
	def get(self, index):
		if index == 0:
			return self.first
		elif index == self._length - 1:
			return self.last
		else:
			current = self._head
			for _ in range(index):
				current = current.next
			return current.data
	
	@first.setter
	def first(self, e):
		if self._length > 0:
			self._head.data = e
		else:
			self.add_first(e)
	
	@last.setter
	def last(self, e):
		if self._length > 0:
			self._tail.data = e
		else:
			self.add_last(e)
	
	def set(self, index, e):
		if index == 0:
			data = self.first
			self.first = e
		elif index == self._length - 1:
			data = self.last
			self.last = e
		else:
			current = self._head
			for _ in range(index):
				current = current.next
			data = current.data
			current.data = e
		return data
	
	def add_first(self, e):
		if self._length == 0:
			self._head = SinglyNode(e)
			self._tail = self._head
		else:
			temp = SinglyNode(e)
			temp.next = self._head
			self._head = temp
		self._length += 1
	
	def add_last(self, e):
		if self._length == 0:
			self._tail = SinglyNode(e)
			self._head = self._tail
		else:
			temp = SinglyNode(e)
			self._tail.next = temp
			self._tail = temp
		self._length += 1
	
	def add(self, index, e):
		if index == 0:
			self.add_last(e)
		elif index == self._length - 1:
			self.add_last(e)
		else:
			previous = self._head
			current = self._head.next
			for _ in range(index):
				previous = current
				current = current.next
			temp = SinglyNode(e)
			previous.next = temp
			temp.next = current
			self._length += 1
	
	def remove_first(self):
		data = self.first
		if self._head == self._tail:
			self.clear()
		else:
			self._head = self._head.next
			self._length -= 1
		return data
	
	def remove_last(self):
		data = self.last
		if self._head == self._tail:
			self.clear()
		else:
			current = self._head
			while current.next != self._tail:
				current = current.next
			current.next = None
			tail = current
			self._length -= 1
		return data
	
	def remove(self, index):
		if index == 0:
			return self.remove_first()
		elif index == self._length - 1:
			return self.remove_last()
		else:
			previous = self._head
			current = self._head.next
			for _ in range(index):
				previous = current
				current = current.next
			previous.next = current.next
			self._length -= 1
			return current.data
	
	def contains(self, e):
		current = self._head
		while current:
			if current.data == e:
				return True
			current = current.next
		return False
	
	def index(self, e):
		current = self._head
		i = 0
		while current:
			if current.data == e:
				return i
			current = current.next
			i += 1
		return -1
	
	def erase(self, e):
		i = self.index(e)
		if i >= 0:
			self.remove(i)
			return True
		return False
	
	def clear(self):
		self._head = None
		self._tail = None
		self._length = 0

File: Ruby/linkedlist/singlylinkedlist.rb
class SinglyNode
   attr_accessor :data, :next
   
   def initialize(data)
		@data = data
		@next = nil
   end
end

class SinglyLinkedList
	attr_reader :size
	
	def initialize(elems = nil)
		@head = nil
		@tail = nil
		@size = 0
		unless elems.nil?
			elems.each { |e| add_last(e) }
		end
	end
	
	def [](index)
		get(index)
	end
	
	def []=(index, e)
		set(index, e)
	end
	
	def length
		@size
	end
	
	def empty?
		@size == 0
	end
	
	def first
		@head&.data
	end
	
	def last
		@tail&.data
	end
	
	def get(index)
		if index == 0
			first
		elsif index == @size - 1
			last
		else
			current = @head
			(index - 1).times { current = current.next }
			current.data
		end
	end
	
	def first=(e)
		if @size > 0
			@head.data = e
		else
			add_first(e)
		end
	end
	
	def last=(e)
		if @size > 0
			@tail.data = e
		else
			add_last(e)
		end
	end
	
	def set(index, e)
		if index == 0
			data = first
			first = e
		elsif index == @size - 1
			data = last
			last = e
		else
			current = @head
			(index - 1).times { current = current.next }
			current.data = e
		end
		data
	end
	
	def add_first(e)
		if @size == 0
			@head = SinglyNode.new(e)
			@tail = @head
		else
			temp = SinglyNode.new(e)
			temp.next = @head
			@head = temp
		end
		@size += 1
	end
	
	def add_last(e)
		if @size == 0
			@tail = SinglyNode.new(e)
			@head = @tail
		else
			temp = SinglyNode.new(e)
			@tail.next = temp
			@tail = temp
		end
		@size += 1
	end
	
	def add(index, e)
		if index == 0
			add_last(e)
		elsif index == @size - 1
			add_last(e)
		else
			previous = @head
			current = @head.next
			(index - 1).times do
				previous = current
				current = current.next
			end
			temp = SinglyNode.new(e)
			previous.next = temp
			temp.next = current
			@size += 1
		end
	end
	
	def remove_first
		data = first
		if @head == @tail
			clear()
		else
			@head = @head.next
			@size -= 1
		end
		data
	end
	
	def remove_last
		data = last
		if @head == @tail
			clear()
		else
			current = @head
			while current != @tail
				current = current.next
			end
			current.next = nil
			@tail = current
			@size -= 1
		end
		data
	end
	
	def remove(index)
		if index == 0
			remove_first()
		elsif index == @size - 1
			remove_last()
		else
			previous = @head
			current = @head.next
			(index - 1).times do
				previous = current
				current = current.next
			end
			previous.next = current.next
			@size -= 1
			current.data
		end
	end
	
	def contains(e)
		current = @head
		while current
			return true if current.data == e
			current = current.next
		end
		false
	end
	
	def index(e)
		current = @head
		i = 0
		while current
			return i if current.data == e
			current = current.next
			i += 1
		end
		-1
	end
	
	def erase(e)
		i = index(e)
		if i >= 0
			remove(i)
			return true
		end
		false
	end
	
	def clear
		@head = nil
		@tail = nil
		@size = 0
	end
end